modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,872 @@
+      The second element is the occurence in the orderedSet, if there's any.
+
+      :arg e: The element to base
+      :type e: `eltType`
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const predecessor(e: eltType): (bool, eltType) {
+      if !isDefaultInitializable(e) {
+        compilerError("predecessor is not avaliable on types that can't be \
+                      default-initialized, here: " + eltType: string);
+      }
+
+      _enter(); defer _leave();
+
+      var result: (bool, eltType);
+      result[0] = false;
+      var baseNode = _find(_root, e);
+      if baseNode == nil {
+        return result;
+      }
+
+      var resultNode = _neighbour(baseNode, 0);
+      if resultNode != nil {
+        result = (true, resultNode!.element);
+      }
+      return result;
+    }
+    
+    /*
+      Find the successor of one element in the orderedSet.
+      Returns if there is such one element.
+      If there is, store the result in `result`.
+
+      Returns a tuple containing two element:
+      The first element is a `bool` that indicates whether there is such an element.
+      The second element is the occurence in the orderedSet, if there's any.
+
+      :arg e: The element to base
+      :type e: `eltType`
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const successor(e: eltType): (bool, eltType) {
+      if !isDefaultInitializable(e) {
+        compilerError("successor is not avaliable on types that can't be \
+                      default-initialized, here: " + eltType: string);
+      }
+
+      _enter(); defer _leave();
+
+      var result: (bool, eltType);
+      result[0] = false;
+      var baseNode = _find(_root, e);
+      if baseNode == nil {
+        return result;
+      }
+
+      var resultNode = _neighbour(baseNode, 1);
+      if resultNode != nil {
+        result = (true, resultNode!.element);
+      }
+      return result;
+    }
+
+    /* Implementation for the `kth` procedure, without acquring the lock */
+    pragma "no doc"
+    proc _kth(node: nodeType, in k: int): nodeType {
+      if node == nil then return nil;
+
+      /* The size of children
+         If nil then 0 ( which is the default value for int )
+      */
+      var childSize: [0..#2] int;
+
+      if node!.children[0] != nil {
+        childSize[0] = node!.children[0]!.size;
+      }
+      if node!.children[1] != nil {
+        childSize[1] = node!.children[1]!.size;
+      }
+
+      // k-th in left subtree
+      if k <= childSize[0] then return _kth(node!.children[0], k);
+      k -= childSize[0];
+
+      // k-th is the current node
+      if k == 1 then return node;
+      k -= 1;
+
+      // k-th is in the right subtree
+      if k <= childSize[1] then return _kth(node!.children[1], k);
+
+      // else k-th is out of range, return nil
+      return nil;
+    }
+
+    /*
+      Find the k-th element in the orderedSet. k starts from 1.
```

Comment:
Few questions here:
- Why starts from 1? Arguably this is internal and not part of the interface, but we are trying to be more 0-based.
- Why `kth` and not just `this`?

Reply:
Maybe the answer to the second question is "because it returns a tuple". If that's the case you can do either:

- Rename it to `get` ?
- Make it return the kth element and not a tuple, and make the users of this check whether k is in bounds/exists via another helper and rename kth to `this`

Reply:
k starts from 1 because here kth indicates the order, not the index. It's natural to say the first element and 0th element sounds awkward. Maybe it's just me.

Reply:
Re 1-based 0-based: We had similar discussions in the recent past. This especially came up when talking about dimensions of a multidimensional array/domain. Today, we have 0th dimension, 1st dimension etc etc. And that is more awkward than your case here. So, as awkward as it may sound, I think we should remain consistent.

Re naming: Chapel has many unique feature that other languages don't have :) That being said, I am not too crazy about `this` either. Is `kth` a common name? I also think that `get` maybe a bit too vague. `getElemAtIdx`? Is it too wordy?

Reply:
@cassella @krishnadey30 -- do you have any opinion here?

Reply:
>  Is kth a common name?

By googling `kth problem`, there are lots of results about kth largest elements or so, which are exactly what `kth` is doing here. 

> `getElemAtIdx`?

My main concern is that using`Idx` could result in confusion between position and order. Terms like index remind me of some kind of arrays and their indexes. Actually, `[]` has the same impression on me. `kth` has no way to do with that.

Reply:
I agree with @Rapiz1 regarding the `kth` being a more intuitive method name, but still, there is also a possibility of having better function names such as `getKthElement`.